engine/src/main/java/org/apache/apex/engine/plugin/DefaultApexPluginDispatcher.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.engine.plugin;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.engine.api.DAGExecutionPluginContext.Handler;
+import org.apache.apex.engine.api.DAGExecutionPluginContext.RegistrationType;
+import org.apache.apex.engine.api.PluginLocator;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.common.util.NameableThreadFactory;
+import com.datatorrent.stram.StramAppContext;
+import com.datatorrent.stram.StreamingContainerManager;
+import com.datatorrent.stram.webapp.AppInfo;
+
+/**
+ * Handle dispatching of events from Stram to Plugins. This implementation creates an executor
+ * service to process the event asynchronously. A separate task {@link DefaultApexPluginDispatcher.ProcessEventTask}
```

Comment:
I incorrectly thought your original comment was on how the plugins were called and I had a concern in that area as well and Tushar had addressed it. I also prefer being conservative and exposing information in a very controlled and thread safe way as well. When I looked at them, all the methods looked thread safe. For example, getOperatorName and getPhysicalOperatoStats, getWindowMetrics refer to thread-safe data structures inside. I  guess the assumption that I made is DAG is unmodifiable and there will be read-only operations, but on retrospect, that is not true as there are methods that can modify it or plan can be changed from outside. It could be abstracted out in a read-only version with a different interface that exposes only the read only methods of the dag and the implementation (LogicalPlan) handles thread safety but that may not ensure a consistent snapshot of the dag (if it is being simulataneously modified from using for example the change logical plan operation). Another option is the snapshot of the dag is taken when the method is called and the snapshot is provided. The taking of the snapshot and changes to logical plan would be synchronized and I am guessing there are only a few places where the changes to the logical plan happen.